Changes to support 2D Decomposition shape changes.  Full I dimension …

GeorgeVandenberghe-NOAA opened the pull request:
…remains in all arrays bu

t SURFCE is altered to use isx:iex bounds which are equivalent to IM which is equivalent to 1:IM

GeorgeVandenberghe-NOAA committed:
Changes to support 2D Decomposition shape changes.  Full I dimension remains in all arrays bu
t SURFCE is altered to use isx:iex bounds which are equivalent to IM which is equivalent to 1:IM

GeorgeVandenberghe-NOAA commented:
Please INSPECT ONLY

GeorgeVandenberghe-NOAA committed:
test of commit only  added one comment to PROCESS.f

GeorgeVandenberghe-NOAA committed:
comment test 2

WenMeng-NOAA commented:
@HuiyaChuang-NOAA, @junwang-noaa, @BoCui-NOAA , @JesseMeng-NOAA 
Hi All,

George filed this PR for inspection only. Please review and provide your comments. Thanks!

WenMeng-NOAA commented on sorc/ncep_post.fd/MPI_FIRST.f:
```diff
@@ -85,8 +87,8 @@ SUBROUTINE MPI_FIRST()
-              nbin_bc, nbin_oc, nbin_su
-
+              nbin_bc, nbin_oc, nbin_su,                                              & 
+              isx,iex
```
@GeorgeVandenberghe-NOAA Would you rename isx, iex as ista, iend which are consistent with jest and jend for j decomposition?

GeorgeVandenberghe-NOAA replied:
ista and iend are already being used in para_range.f.  The conflict can be resolved because they aren't used outside
but I will have to use other names inside this code

WenMeng-NOAA commented on sorc/ncep_post.fd/MPI_FIRST.f:
```diff
@@ -149,7 +157,8 @@ SUBROUTINE MPI_FIRST()
       do i = 0, num_procs - 1
-         call para_range(1,jm,num_procs,i,jsx,jex) 
+         call para_range2(1,jm,1,im,num_procs,i,jsx,jex,isx,iex) 
+!gwv  delete after 2D support is validated        call para_range(1,jm,num_procs,i,jsx,jex) 
```
@GeorgeVandenberghe-NOAA Would you consider to assign i halo for ista_2l, jend_2u, ivend_2u?

junwang-noaa replied:
@GeorgeVandenberghe-NOAA do you want to have num_procs in i and j dimensions separately so that we will have some flexibility to control mpi tasks assigned to i and j dimension? E.g. we can have num_procs_i, num_procs_j, and num_procs=num_procs_i * num_procs_j. We have still use para_range for i dimension with num_procs_i tasks.

HuiyaChuang-NOAA replied:
> @GeorgeVandenberghe-NOAA do you want to have num_procs in i and j dimensions separately so that we will have some flexibility to control mpi tasks assigned to i and j dimension? E.g. we can have num_procs_i, num_procs_j, and num_procs=num_procs_i * num_procs_j. We have still use para_range for i dimension with num_procs_i tasks.

is this how it's done in FV3 model?

HuiyaChuang-NOAA replied:
> @GeorgeVandenberghe-NOAA do you want to have num_procs in i and j dimensions separately so that we will have some flexibility to control mpi tasks assigned to i and j dimension? E.g. we can have num_procs_i, num_procs_j, and num_procs=num_procs_i * num_procs_j. We have still use para_range for i dimension with num_procs_i tasks.

I think this is a great idea if this can be done.  It gives us flexibility to go back to decompose Y only if necessary

WenMeng-NOAA commented on sorc/ncep_post.fd/TIMEF.f:
```diff
@@ -3,17 +3,20 @@
 !mp     HPs)  Designed to duplicate timef (elapsed time in milliseconds)
 !
         function timef()
```
@GeorgeVandenberghe-NOAA There are timer updates in UPP code recently. Would these timer updates be used for your testing only or other intentions?

GeorgeVandenberghe-NOAA replied:
Additional timer codes are only for my testing.  But TIMEF.f  itself is not producing valid times on some systems and I fixed that in passing.   MPI_WTIME provides the highest resolution timer on our intel based systems; much better than rtc()

junwang-noaa commented on sorc/ncep_post.fd/PARA_RANGE.f:
```diff
@@ -46,4 +46,22 @@ SUBROUTINE PARA_RANGE (N1,N2,NPROCS,IRANK,ISTA,IEND)
+      iend   = ista + iwork1 - 1
+      if ( iwork2 > irank ) iend = iend + 1
+      isx=i1
+      iex=i2
```
@GeorgeVandenberghe-NOAA Can you explain how this two dimension decomposition method is working? I do not see the i dimension decomposition.

BoCui-NOAA replied:
I guess only the option of the two dimension decomposition is setup.  So the i direction is still from 1 to im.

GeorgeVandenberghe-NOAA commented:
*It's not working yet*.  I am adding the support for it in array shapes and
changing loop boundaries  first before actually processing subsets of the I
dimension on different processors. right now the subdomain  *isx:iex *is
the full *1:im* domain.  I'll rename isx and iex as mentioned in another
mail message.

The most difficult part of this effort (besides debugging halos and the one
or two places where something needs to be full  domain and I missed it) is
getting the model forecast (what I call state though it's not
mathematically just state).   out to all of the processors.   With a 1D
decomposition we can do a scatter or MPI_IO
reading one subset of a record on each task.  The scatter is more (just
bookkeeping but more) complicated with a 2D decomposition
and gathers for grib2 writing will also be a little tougher.


thanks all for reviewing.  This was submitted for design comments.


On Thu, Apr 8, 2021 at 10:14 AM Jun Wang ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/PARA_RANGE.f
> <https://github.com/NOAA-EMC/EMC_post/pull/285#discussion_r609747075>:
>
> > @@ -46,4 +46,22 @@ SUBROUTINE PARA_RANGE (N1,N2,NPROCS,IRANK,ISTA,IEND)
>        if ( iwork2 > irank ) iend = iend + 1
>        return
>        end
> +      SUBROUTINE PARA_RANGE2 (N1,N2,i1,i2,NPROCS,IRANK,ISTA,IEND,isx,iex)
> +
> +      implicit none
> +      integer,intent(in)  ::  n1,n2,nprocs,irank,i1,i2
> +      integer,intent(out) ::  ista,iend,isx,iex
> +      integer iwork1, iwork2
> +
> +      iwork1 = ( n2 - n1 + 1 ) / nprocs
> +      iwork2 = mod ( n2 - n1 + 1, nprocs )
> +      ista   = irank * iwork1 + n1 + min ( irank, iwork2 )
> +      iend   = ista + iwork1 - 1
> +      if ( iwork2 > irank ) iend = iend + 1
> +      isx=i1
> +      iex=i2
>
> @GeorgeVandenberghe-NOAA <https://github.com/GeorgeVandenberghe-NOAA> Can
> you explain how this two dimension decomposition method is working? I do
> not see the i dimension decomposition.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/285#pullrequestreview-631448255>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FQA26M6265ABAFRAF3THW22NANCNFSM42HRRUAA>
> .
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

GeorgeVandenberghe-NOAA commented:
Does anyone have a set of grib record by record tests of the post output.
cmp is not a valid test because
there seems to be occasionall grib padding which changes the file without
changing the actual record contents
so I need a record by record compare test when checking for regressions

On Thu, Apr 8, 2021 at 10:38 AM George Vandenberghe - NOAA Affiliate <
***@***.***> wrote:

> *It's not working yet*.  I am adding the support for it in array shapes
> and changing loop boundaries  first before actually processing subsets of
> the I dimension on different processors. right now the subdomain  *isx:iex
> *is the full *1:im* domain.  I'll rename isx and iex as mentioned in
> another mail message.
>
> The most difficult part of this effort (besides debugging halos and the
> one or two places where something needs to be full  domain and I missed it)
> is
> getting the model forecast (what I call state though it's not
> mathematically just state).   out to all of the processors.   With a 1D
> decomposition we can do a scatter or MPI_IO
> reading one subset of a record on each task.  The scatter is more (just
> bookkeeping but more) complicated with a 2D decomposition
> and gathers for grib2 writing will also be a little tougher.
>
>
> thanks all for reviewing.  This was submitted for design comments.
>
>
> On Thu, Apr 8, 2021 at 10:14 AM Jun Wang ***@***.***> wrote:
>
>> ***@***.**** commented on this pull request.
>> ------------------------------
>>
>> In sorc/ncep_post.fd/PARA_RANGE.f
>> <https://github.com/NOAA-EMC/EMC_post/pull/285#discussion_r609747075>:
>>
>> > @@ -46,4 +46,22 @@ SUBROUTINE PARA_RANGE (N1,N2,NPROCS,IRANK,ISTA,IEND)
>>        if ( iwork2 > irank ) iend = iend + 1
>>        return
>>        end
>> +      SUBROUTINE PARA_RANGE2 (N1,N2,i1,i2,NPROCS,IRANK,ISTA,IEND,isx,iex)
>> +
>> +      implicit none
>> +      integer,intent(in)  ::  n1,n2,nprocs,irank,i1,i2
>> +      integer,intent(out) ::  ista,iend,isx,iex
>> +      integer iwork1, iwork2
>> +
>> +      iwork1 = ( n2 - n1 + 1 ) / nprocs
>> +      iwork2 = mod ( n2 - n1 + 1, nprocs )
>> +      ista   = irank * iwork1 + n1 + min ( irank, iwork2 )
>> +      iend   = ista + iwork1 - 1
>> +      if ( iwork2 > irank ) iend = iend + 1
>> +      isx=i1
>> +      iex=i2
>>
>> @GeorgeVandenberghe-NOAA <https://github.com/GeorgeVandenberghe-NOAA>
>> Can you explain how this two dimension decomposition method is working? I
>> do not see the i dimension decomposition.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/NOAA-EMC/EMC_post/pull/285#pullrequestreview-631448255>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ANDS4FQA26M6265ABAFRAF3THW22NANCNFSM42HRRUAA>
>> .
>>
>
>
> --
>
> George W Vandenberghe
>
> *IMSG* at NOAA/NWS/NCEP/EMC
>
> 5830 University Research Ct., Rm. 2141
>
> College Park, MD 20740
>
> ***@***.***
>
> 301-683-3769(work)  3017751547(cell)
>
>

-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

WenMeng-NOAA commented:
@GeorgeVandenberghe-NOAA @GeorgeVandenberghe-NOAA The baselines of UPP regression tests on Hera at
/scratch2/NCEPDEV/ovp/Wen.Meng/ncep_post/post_regression_test_new/data_out

You may do field by field comparison for two grib2 files as:
/home/Wen.Meng/bin/cmp_grib2_grib2_new file1 file2

GeorgeVandenberghe-NOAA commented:
Thanks.

On Fri, Apr 9, 2021 at 3:57 PM WenMeng-NOAA ***@***.***>
wrote:

> @GeorgeVandenberghe-NOAA <https://github.com/GeorgeVandenberghe-NOAA>
> @GeorgeVandenberghe-NOAA <https://github.com/GeorgeVandenberghe-NOAA> The
> baselines of UPP regression tests on Hera at
> /scratch2/NCEPDEV/ovp/Wen.Meng/ncep_post/post_regression_test_new/data_out
>
> You may do field by field comparison for two grib2 files as:
> /home/Wen.Meng/bin/cmp_grib2_grib2_new file1 file2
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/285#issuecomment-816933431>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FVAEXSYY6JIO6EXBTDTH5LY5ANCNFSM42HRRUAA>
> .
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

GeorgeVandenberghe-NOAA commented:
Yes this is how it's done in fv3 which I believe is now supposed to be
called UFS.

On Fri, Apr 9, 2021 at 4:28 PM HuiyaChuang-NOAA ***@***.***>
wrote:

> *@HuiyaChuang-NOAA* commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/MPI_FIRST.f
> <https://github.com/NOAA-EMC/EMC_post/pull/285#discussion_r610888329>:
>
> > @@ -149,7 +157,8 @@ SUBROUTINE MPI_FIRST()
>  !     counts, disps for gatherv and scatterv
>  !
>        do i = 0, num_procs - 1
> -         call para_range(1,jm,num_procs,i,jsx,jex)
> +         call para_range2(1,jm,1,im,num_procs,i,jsx,jex,isx,iex)
> +!gwv  delete after 2D support is validated        call para_range(1,jm,num_procs,i,jsx,jex)
>
> @GeorgeVandenberghe-NOAA <https://github.com/GeorgeVandenberghe-NOAA> do
> you want to have num_procs in i and j dimensions separately so that we will
> have some flexibility to control mpi tasks assigned to i and j dimension?
> E.g. we can have num_procs_i, num_procs_j, and num_procs=num_procs_i *
> num_procs_j. We have still use para_range for i dimension with num_procs_i
> tasks.
>
> is this how it's done in FV3 model?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/285#discussion_r610888329>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FUXCXA4AE5BWUTQICDTH5PPTANCNFSM42HRRUAA>
> .
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

HuiyaChuang-NOAA reviewed (commented):
@GeorgeVandenberghe-NOAA @junwang-noaa 
I think your changes looked fine,  I do like Jun's suggestions to decompose i,j separately and you mentioned this is how it's done in UFS.  If you could move toward that direction, it would be great.  
@GeorgeVandenberghe-NOAA @WenMeng-NOAA @BoCui-NOAA @JesseMeng-NOAA 
Also, my thought is for you to modify mpi_first.f, para_range.f, process.f wrfpost.f, and perhaps mdlfld and mdl2p (these two are first to be called by process.f) and run a test by making other subroutines called in process.f stubs and see if things are working properly.  If so, then I'm hoping Bo and Jesse will take over modifying the rest of UPP subroutines.

GeorgeVandenberghe-NOAA committed:
Added more routines to 2D decomposition

GeorgeVandenberghe-NOAA commented:
In changing isx to ISTA as requested, I find ISTA is used in MDL2P.f for another purpose.  I have renamed the local ISTA in MDL2P.f
Otherwise I have gone through and changed all isx and iex to ISTA and IEND as requested and it still compiles and reproduces.  Note 2D subsetting is NOT implemented yet.  I am just laying the structure for it.   IEND is also used in PARA_CONFIG.f to set the J boundary.   I changed that.

GeorgeVandenberghe-NOAA committed:
Changed isx and iex to ISTA and IEND in listed routines.   ISTA had prior use in MDL2P.f and that needed changing.
IEND had prior use in PARA_CONFIG, changed to IENDJ in PARA_CONFIG2 where needed PARA_CONFIG did not
need changing

GeorgeVandenberghe-NOAA commented:
THe renamed start and end bounds ista and iend are now in the PR

Commit:
Merge remote-tracking branch 'upstream/develop' into develop

Conflicts:
	sorc/ncep_post.fd/CLDRAD.f
	sorc/ncep_post.fd/CTLBLK.f
	sorc/ncep_post.fd/MDL2AGL.f
	sorc/ncep_post.fd/MDL2P.f
	sorc/ncep_post.fd/MDL2SIGMA2.f
	sorc/ncep_post.fd/MDLFLD.f
	sorc/ncep_post.fd/MISCLN.f
	sorc/ncep_post.fd/SURFCE.f
	sorc/ncep_post.fd/TIMEF.f
	sorc/ncep_post.fd/WRFPOST.f

Commit:
test commit of new base into clone

Commit:
repair of a bunch of changed routines somehow corrupted by a merge 5/3/2021

GeorgeVandenberghe-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into develop

GeorgeVandenberghe-NOAA committed:
fixed 5/25 problem

GeorgeVandenberghe-NOAA committed:
added ctlblk change to support ista and iend partial I dimensons

GeorgeVandenberghe-NOAA committed:
added para_range2 to PARA_RANGE.f to support 2D decomposition

GeorgeVandenberghe-NOAA committed:
Modified MPI_FIRST.f to support a 2D decomposition but the actual numbers used
remain the 1D special case and the changes just add two more indices for start and end I domains
and logic to convert scatter counts and displacements to the product of the I and J subdomain sizes rather than I full domain x J subdomain.

GeorgeVandenberghe-NOAA commented:
Added support for a 2D decomposition (externally supplied)  in MPI_FIRST.f   MPI_FIRST is configured for the special case where the I dimension is decomposed over the full 1:IM domain but para_config2  which replaces para_config, replaces IM with ista:iend to specify the I range and the scatter displacement and count arrays are sized for 2D cells rather than 1D slices

HuiyaChuang-NOAA commented on sorc/ncep_post.fd/MPI_FIRST.f:
```diff
@@ -116,7 +119,18 @@ SUBROUTINE MPI_FIRST()
+!  subdivided the third and fourth arguments will have to be integral
+!  factors of num_procs and on 5/27/21 I am still working out a general
+!  way to do this if the user doesn't select the factors
+      call para_range2(im,jm,1,num_procs,me,ista,iend,jsta,jend)
```
@GeorgeVandenberghe-NOAA this is good as a control case to match what UPP is doing now

GeorgeVandenberghe-NOAA replied:
Yes.  It allows UPP to function in the 1D decomposition with the 2D decomposition support gradually inserted.  The big change and test will be when the third argument of para_range2 is larger than 1 which can't be done until all arrays  are dimensioned  with the  I subdomain.   Easy for MOST but I need to work out the halo stuff which both you and wen caught in the para_range2 evaluaton.

HuiyaChuang-NOAA commented on sorc/ncep_post.fd/MPI_FIRST.f:
```diff
@@ -116,7 +119,18 @@ SUBROUTINE MPI_FIRST()
+!  factors of num_procs and on 5/27/21 I am still working out a general
+!  way to do this if the user doesn't select the factors
+      call para_range2(im,jm,1,num_procs,me,ista,iend,jsta,jend)
       jsta_m  = jsta
```
Will you add ista_m, ista_m2, etc later on?

GeorgeVandenberghe-NOAA commented:
Yes.  I will add them.   Havent worked out what they should be yet.


On Fri, May 28, 2021 at 2:10 PM HuiyaChuang-NOAA ***@***.***>
wrote:

> *@HuiyaChuang-NOAA* commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/MPI_FIRST.f
> <https://github.com/NOAA-EMC/EMC_post/pull/285#discussion_r641728052>:
>
> > @@ -116,7 +119,18 @@ SUBROUTINE MPI_FIRST()
>  !
>  !     global loop ranges
>  !
> -      call para_range(1,jm,num_procs,me,jsta,jend)
> +!      call para_range(1,jm,num_procs,me,jsta,jend)
> +! GWVX temporary documentation
> +!  para_range2 supports a 2D decomposition.  The rest of the post
> +!  supports 1D still and the call here is the special case where each
> +!  processor gets all of the longitudes in the latitude 1D subdomain
> +!  jsta:jend.  The X decomposition will be specified by the third
> +!  argument (currently 1) and the Y decoposition will be specified by
> +!  the fourth argument (currently all of the ranks)   When X is
> +!  subdivided the third and fourth arguments will have to be integral
> +!  factors of num_procs and on 5/27/21 I am still working out a general
> +!  way to do this if the user doesn't select the factors
> +      call para_range2(im,jm,1,num_procs,me,ista,iend,jsta,jend)
>        jsta_m  = jsta
>
> Will you add ista_m, ista_m2, etc later on?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/285#pullrequestreview-671539919>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FXIB2B4PWNGGDRE6ODTP7MBFANCNFSM42HRRUAA>
> .
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

WenMeng-NOAA commented on sorc/ncep_post.fd/CTLBLK.f:
```diff
@@ -54,7 +54,7 @@ module CTLBLK_mod
-  integer :: NUM_PROCS,ME,JSTA,JEND,JSTA_M,JEND_M,                     &
+  integer :: NUM_PROCS,ME,JSTA,JEND,ista,iend,JSTA_M,JEND_M,                     &
              JSTA_M2,JEND_M2,IUP,IDN,ICNT(0:1023),IDSP(0:1023),        &
              JSTA_2L, JEND_2U,JVEND_2u,NUM_SERVERS, MPI_COMM_INTER,    &
```
@GeorgeVandenberghe-NOAA Will you consider adding ISTA_2L and IDEN_2u for x decomposition?

GeorgeVandenberghe-NOAA commented:
Yes.  This was a miss.

Do you know how the jend+1:jend+2 range of arrays dimensioned that way, is
populated.  The scatter doesn't do it.


On Fri, May 28, 2021 at 3:10 PM WenMeng-NOAA ***@***.***>
wrote:

> *@WenMeng-NOAA* commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/CTLBLK.f
> <https://github.com/NOAA-EMC/EMC_post/pull/285#discussion_r641765924>:
>
> > @@ -54,7 +54,7 @@ module CTLBLK_mod
>            SPL(komax),ALSL(komax),PREC_ACC_DT,PT_TBL,PREC_ACC_DT1,spval
>  ! real :: SPVAL=9.9e10                                     ! Moorthi
>  !
> -  integer :: NUM_PROCS,ME,JSTA,JEND,JSTA_M,JEND_M,                     &
> +  integer :: NUM_PROCS,ME,JSTA,JEND,ista,iend,JSTA_M,JEND_M,                     &
>               JSTA_M2,JEND_M2,IUP,IDN,ICNT(0:1023),IDSP(0:1023),        &
>               JSTA_2L, JEND_2U,JVEND_2u,NUM_SERVERS, MPI_COMM_INTER,    &
>
> @GeorgeVandenberghe-NOAA <https://github.com/GeorgeVandenberghe-NOAA>
> Will you consider adding ISTA_2L and IDEN_2u for x decomposition?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/285#pullrequestreview-671579045>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FU5IDBW4TEND22OXFTTP7TBDANCNFSM42HRRUAA>
> .
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

BoCui-NOAA commented on sorc/ncep_post.fd/PARA_RANGE.f:
```diff
@@ -46,4 +46,42 @@ SUBROUTINE PARA_RANGE (N1,N2,NPROCS,IRANK,ISTA,IEND)
       return
       end
+!!
+!! USAGE:    CALL PARA_RANGE2(N1,N2,NPROCS,IRANK,ISTA,IEND)(A)
```
There are 9 input arguments for para_range2,  I suggest correct this line to:

!! USAGE:    CALL PARA_RANGE2(N1,N2,NX,NY,NRANK,ISTA,IEND)

This would be consistent with the below description.

GeorgeVandenberghe-NOAA commented:
OOPS.  HARD MISS!!.  I'll fix this this afternoon.

On Fri, May 28, 2021 at 3:50 PM BoCui-NOAA ***@***.***> wrote:

> *@BoCui-NOAA* commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/PARA_RANGE.f
> <https://github.com/NOAA-EMC/EMC_post/pull/285#discussion_r641784057>:
>
> > @@ -46,4 +46,42 @@ SUBROUTINE PARA_RANGE (N1,N2,NPROCS,IRANK,ISTA,IEND)
>        if ( iwork2 > irank ) iend = iend + 1
>        return
>        end
> +!!
> +!! USAGE:    CALL PARA_RANGE2(N1,N2,NPROCS,IRANK,ISTA,IEND)(A)
>
> There are 9 input arguments for para_range2, I suggest correct this line
> to:
>
> !! USAGE: CALL PARA_RANGE2(N1,N2,NX,NY,NRANK,ISTA,IEND)
>
> This would be consistent with the below description.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/285#pullrequestreview-671601912>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FUPDFQ4C6UJVMUKFBDTP7XXXANCNFSM42HRRUAA>
> .
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

BoCui-NOAA commented on sorc/ncep_post.fd/PARA_RANGE.f:
```diff
@@ -46,4 +46,42 @@ SUBROUTINE PARA_RANGE (N1,N2,NPROCS,IRANK,ISTA,IEND)
+     subroutine para_range2(im,jm,nx,ny,nrank,ista,iend,jsta,jend)
+         jx=nrank/nx
+         ix=nrank-(jx*nx)
+           call para_range(1,im,nx,ix,ista,iend)
```
Here ix and jx are task id,  i.e irank in subroutine para_range.  If so, I suggest give them some other names like me_x, me_y, or irank/jrank. How do you think?

For this case, ix is set as 0, correct?  

Thanks!

GeorgeVandenberghe-NOAA commented:
Yes that would probably be clearer.  If this PR gets committed I'll make
these changes in the next one along with addition of iend_2u,iend_2l
and the other halo bounded im settings and the documentation fix.    If it
doesn't get committed I'll make these changes this afternoon.


On Fri, May 28, 2021 at 4:13 PM BoCui-NOAA ***@***.***> wrote:

> *@BoCui-NOAA* commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/PARA_RANGE.f
> <https://github.com/NOAA-EMC/EMC_post/pull/285#discussion_r641793870>:
>
> > +!!   OUTPUT FILES:
> +!!     STDOUT  - RUN TIME STANDARD OUT.
> +!!
> +!!   SUBPROGRAMS CALLED:
> +!!     UTILITIES:
> +!!       NONE
> +!!     LIBRARY:
> +!!
> +!!   ATTRIBUTES:
> +!!     LANGUAGE: FORTRAN
> +!!     MACHINE : IBM RS/6000 SP
> +!!
> +     subroutine para_range2(im,jm,nx,ny,nrank,ista,iend,jsta,jend)
> +         jx=nrank/nx
> +         ix=nrank-(jx*nx)
> +           call para_range(1,im,nx,ix,ista,iend)
>
> Here ix and jx are task id, i.e irank in subroutine para_range. If so, I
> suggest give them some other names like me_x, me_y, or irank/jrank. How do
> you think?
>
> For this case, ix is set as 0, correct?
>
> Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/285#pullrequestreview-671614506>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FREXIBWFMT6KLTIE2TTP72MRANCNFSM42HRRUAA>
> .
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

GeorgeVandenberghe-NOAA committed:
Merge remote-tracking branch 'upstream/develop' into develop

GeorgeVandenberghe-NOAA committed:
Added support for halo settings of 2D boundaries i.e ista_2l to match the 1D analogs long present in the source
6/1/2021

GeorgeVandenberghe-NOAA commented:
I have added support for 2D decomposition left  and right boundary neighbor setting to match those settings in the 1D decompostion, in MPI_FIRST.f (used in module only)
and CTLBLK.f (defined and available for everyone else in the module).  I pushed it to my clone and the PR should now be automatically updated with those changes.
Wen commented on their absence.

GeorgeVandenberghe-NOAA committed:
Added definitions for ista_2u, and ista_2u in MPI_FIRST.f

GeorgeVandenberghe-NOAA committed:
Reshaped arrays  in MDL2FLD.f to support 2D decomposition.  Modified TWO (only) loops to try thie

BoCui-NOAA commented on sorc/ncep_post.fd/MPI_FIRST.f:
```diff
@@ -82,20 +82,25 @@ SUBROUTINE MPI_FIRST()
-              nbin_bc, nbin_oc, nbin_su
+              nbin_bc, nbin_oc, nbin_su,                                               &
+              ISTA_M,IEND_M,ISTA_M2,IEND_M2,                                          &
+              ISTA_2L, IEND_2U,IVEND_2U             
```
I suggest use lowcase for all variables, which looks more consistent.

ISTA_M,IEND_M,ISTA_M2,IEND_M2, 
ISTA_2L, IEND_2U,IVEND_2U 

replaced to:
ista_m,iend_m,ista_m2,iend_m2,
ista_2l, iend_2u,ivend_2u





    
    
      
    <span class="blob-code-inner blob-code-marker" data-code-marker="+">              ISTA_2L, IEND_2U,IVEND_2U             </span><!--EndFragment-->

GeorgeVandenberghe-NOAA commented:
Will fix in the next update.

On Thu, Jun 3, 2021 at 1:24 PM BoCui-NOAA ***@***.***> wrote:

> *@BoCui-NOAA* commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/MPI_FIRST.f
> <https://github.com/NOAA-EMC/EMC_post/pull/285#discussion_r644991391>:
>
> >                jend_m, jend_m2, iup, idn, icnt, im, idsp, jsta_2l, jend_2u,            &
>                jvend_2u, lm, lp1, jsta_2l, jend_2u, nsoil, nbin_du, nbin_ss,           &
> -              nbin_bc, nbin_oc, nbin_su
> +              nbin_bc, nbin_oc, nbin_su,                                               &
> +              ISTA_M,IEND_M,ISTA_M2,IEND_M2,                                          &
> +              ISTA_2L, IEND_2U,IVEND_2U
>
> I suggest use lowcase for all variables, which looks more consistent.
>
> ISTA_M,IEND_M,ISTA_M2,IEND_M2,
> ISTA_2L, IEND_2U,IVEND_2U
>
> replaced to:
> ista_m,iend_m,ista_m2,iend_m2,
> ista_2l, iend_2u,ivend_2u
>
> <span class="blob-code-inner blob-code-marker" data-code-marker="+">              ISTA_2L, IEND_2U,IVEND_2U             </span><!--EndFragment-->
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/285#pullrequestreview-675547517>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FSDHD76WKYATYQHJL3TQ63GJANCNFSM42HRRUAA>
> .
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

HuiyaChuang-NOAA commented on sorc/ncep_post.fd/PARA_RANGE.f:
```diff
@@ -46,4 +46,42 @@ SUBROUTINE PARA_RANGE (N1,N2,NPROCS,IRANK,ISTA,IEND)
+!!   INPUT ARGUMENT LIST:
+!!     N1 - LAAT INTERATE VALUE I dimension
+!!     N2 - LAST INTERATE VALUE  J dimension
+!!     NX  NUMBER OF subdomains in Z  dimension
```
NUMBER OF subdomains in X  dimension

HuiyaChuang-NOAA commented on sorc/ncep_post.fd/PARA_RANGE.f:
```diff
@@ -46,4 +46,42 @@ SUBROUTINE PARA_RANGE (N1,N2,NPROCS,IRANK,ISTA,IEND)
+!!   INPUT ARGUMENT LIST:
+!!     N1 - LAAT INTERATE VALUE I dimension
+!!     N2 - LAST INTERATE VALUE  J dimension
+!!     NX  NUMBER OF subdomains in Z  dimension
```
in X dimension

HuiyaChuang-NOAA replied:
@GeorgeVandenberghe-NOAA Checked the updated version of para_range.f, mpi_first.f and ctlblk.f and they look fine.

BoCui-NOAA commented on sorc/ncep_post.fd/PARA_RANGE.f:
```diff
@@ -46,4 +46,42 @@ SUBROUTINE PARA_RANGE (N1,N2,NPROCS,IRANK,ISTA,IEND)
+!!
+     subroutine para_range2(im,jm,nx,ny,nrank,ista,iend,jsta,jend)
+         jx=nrank/nx
+         ix=nrank-(jx*nx)
```
ix is always set as 0 in this mode, will it be flexible for future development?

GeorgeVandenberghe-NOAA commented:
Typo

On Thu, Jun 3, 2021 at 2:23 PM HuiyaChuang-NOAA ***@***.***>
wrote:

> *@HuiyaChuang-NOAA* commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/PARA_RANGE.f
> <https://github.com/NOAA-EMC/EMC_post/pull/285#discussion_r645031505>:
>
> > @@ -46,4 +46,42 @@ SUBROUTINE PARA_RANGE (N1,N2,NPROCS,IRANK,ISTA,IEND)
>        if ( iwork2 > irank ) iend = iend + 1
>        return
>        end
> +!!
> +!! USAGE:    CALL PARA_RANGE2(N1,N2,NX,NY,NRANK,ISTA,IEND,JSTA,JEND)(A)
> +!!   INPUT ARGUMENT LIST:
> +!!     N1 - LAAT INTERATE VALUE I dimension
> +!!     N2 - LAST INTERATE VALUE  J dimension
> +!!     NX  NUMBER OF subdomains in Z  dimension
>
> in X dimension
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/285#pullrequestreview-675601736>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FV4OECPVFDWVZBHMETTQ7CCXANCNFSM42HRRUAA>
> .
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

GeorgeVandenberghe-NOAA commented:
Yes it will be flexible as soon as 2D is activated by partial I range
settings.


On Thu, Jun 3, 2021 at 2:28 PM BoCui-NOAA ***@***.***> wrote:

> *@BoCui-NOAA* commented on this pull request.
> ------------------------------
>
> In sorc/ncep_post.fd/PARA_RANGE.f
> <https://github.com/NOAA-EMC/EMC_post/pull/285#discussion_r645034641>:
>
> > +!!
> +!!   OUTPUT FILES:
> +!!     STDOUT  - RUN TIME STANDARD OUT.
> +!!
> +!!   SUBPROGRAMS CALLED:
> +!!     UTILITIES:
> +!!       NONE
> +!!     LIBRARY:
> +!!
> +!!   ATTRIBUTES:
> +!!     LANGUAGE: FORTRAN
> +!!     MACHINE : IBM RS/6000 SP
> +!!
> +     subroutine para_range2(im,jm,nx,ny,nrank,ista,iend,jsta,jend)
> +         jx=nrank/nx
> +         ix=nrank-(jx*nx)
>
> ix is always set as 0 in this mode, will it be flexible for future
> development?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/EMC_post/pull/285#pullrequestreview-675605634>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANDS4FRB3H5DL5ZDQ4E2E5LTQ7CU7ANCNFSM42HRRUAA>
> .
>


-- 

George W Vandenberghe

*IMSG* at NOAA/NWS/NCEP/EMC

5830 University Research Ct., Rm. 2141

College Park, MD 20740

***@***.***

301-683-3769(work)  3017751547(cell)

GeorgeVandenberghe-NOAA committed:
sync development head with gv fork 8/2/2021

Merge remote-tracking branch 'upstream/develop' into develop

WenMeng-NOAA commented:
@GeorgeVandenberghe-NOAA Now all development of 2d decomposition were transitioned to the PR #339. Could you withdraw this PR? Thanks!
